### PR TITLE
add chain to Set (Fixes #344)

### LINKED ISSUE #344: add 'chain' to Set

### PR DESCRIPTION
**Pull request by sledorze**

### CODE REVIEW

File: src/Set.ts [hunk 1]
```diff
@@ -36,6 +36,16 @@ export const every = <A>(x: Set<A>, predicate: Predicate<A>): boolean => {
   return !some(x, not(predicate))
 }
 
+/** @function */
+export const chain = <A, B>(x: Set<A>, bset: Setoid<B>, f: (x: A) => Set<B>): Set<B> => {
```
Comment (gcanti) [inline src/Set.ts]: Same as `map`, shouldn't be

```ts
export const chain = <B>(S: Setoid<B>) => <A>(x: Set<A>, f: (x: A) => Set<B>): Set<B>
```

?

Comment (sledorze) [reply src/Set.ts]: Ok Done

File: src/Set.ts [hunk 2]
```diff
@@ -36,6 +36,16 @@ export const every = <A>(x: Set<A>, predicate: Predicate<A>): boolean => {
   return !some(x, not(predicate))
 }
 
+/** @function */
+export const chain = <B>(bset: Setoid<B>) => <A>(x: Set<A>, f: (x: A) => Set<B>): Set<B> => {
+  let r = new Set<B>()
+  const setUnion = union(bset)
+  x.forEach(e => {
+    r = setUnion(f(e), r)
```
Comment (gcanti) [inline src/Set.ts]: I wonder, shouldn't we use mutability here for performance reasons?

Comment (sledorze) [reply src/Set.ts]: Done

**Pull request merged**